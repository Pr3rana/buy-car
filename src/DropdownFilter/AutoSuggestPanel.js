import React from 'react';
import './DropdownFilter.css';
// list: [{label, id}]

// function AutoSuggestPanel({filterData, select, type}){
  
  
//   return (
//       // <div  className="custom-select" key={type}>
//           /* <p>{type}</p> */
//           /* <div ref= {typeRefs[`default${type}`]} className="default-selected-item" onClick={()=>typeRefs[type].current.style.display!==""?typeRefs[type].current.style.display="":typeRefs[type].current.style.display="block"}>All {type}</div> */
//           <SuggestionRow ref={typeRefs[type]} list={filterData} onClick={select} type={type}/>
//       // </div>
//   )
// }

function AutoSuggestPanel ({filterData=[], select, type}, ref) {
  // const [defaultSelect, setDefaultSelect] = useState(`All ${type}`);
  // const typeRefs = {};
  // typeRefs[type] = useRef(null);
  // typeRefs[`default${type}`] = useRef(null);
    return(
      <div ref={ref} className="autosuggestPanel">
        {
          /*Mapping the relevancy search result to  autosuggest*/
          filterData.map((el, index) => {
          return (
              <div className="suggestionRow" key = {el+index} onClick={(e)=>select(e)}  data-filter-type={type}>{el}</div>
          )
        })
        }
      </div>
    )
  }
  const typeRef = React.forwardRef(AutoSuggestPanel);
  export default typeRef;