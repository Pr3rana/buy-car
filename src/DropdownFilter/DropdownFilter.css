.custom-select {
    position: relative;
    font-family: Arial;
  }
  .default-selected-item{
    border: 1px solid #ededed;
    padding: 12px;
    width: 84%;
  }
  .default-selected-item.select-arrow-active:after {
    border-color: transparent transparent #ededed transparent;
    top: 7px;
  }

  .default-selected-item:after {
    position: absolute;
    content: "";
    top: 52px;
    right: 60px;
    width: 0;
    height: 0;
    border: 8px solid #ededed;
    border-color: #ededed transparent transparent transparent;
  }

  .autosuggestPanel{
    margin-top: 12px;
    width: 84%;
    border: 1px solid #ededed;
    border-bottom: none;
    z-index: 99;
    top: 100%;
    left: 0;
    right: 0;
    overflow-y: auto;
    max-height: 50vh;
    display: none;
  }
  
  .suggestionRow {
    padding: 10px;
    cursor: pointer;
    background-color: #fff; 
    border-bottom: 1px solid #d4d4d4; 
  }
  
  /*when hovering an item:*/
  .suggestionRow:hover {
    background-color: #ea7f28;
    color: #ededed;
  }
