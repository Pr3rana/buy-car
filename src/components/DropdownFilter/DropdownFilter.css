.custom-select {
  font-size: 18px;
  position: relative;
  font-family: Arial;
}
.default-selected-item{
  border: 1px solid #ededed;
  padding: 12px;
}
.default-selected-item.select-arrow-active:after {
  border-color: transparent transparent #ededed transparent;
  top: 41px;
}

.filter-type{
  font-size: 14px;
}

.default-selected-item:after {
  position: absolute;
  content: "";
  top: 52px;
  right: 20px;
  width: 0;
  height: 0;
  border: 8px solid #ededed;
  border-color: #ededed transparent transparent transparent;
}

.autosuggestPanel{
  margin-top: 12px;
  border: 1px solid #ededed;
  border-bottom: none;
  overflow-y: auto;
  max-height: 20vh;
  display: none;
}

.filter-loader{
  text-align: center;
}
.filter-loader img{
  width: 50px;
  height: 50px;
}
  
.suggestionRow {
  padding: 10px;
  cursor: pointer;
  background-color: #fff; 
  border-bottom: 1px solid #d4d4d4; 
}
  
.suggestionRow:hover {
  background-color: #ea7f28;
  color: #ededed;
}

.filter-btn{
  font-size: 18px;
  margin: 18px 0 18px 18px;
}