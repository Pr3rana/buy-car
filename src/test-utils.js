import React from 'react';
import { render } from '@testing-library/react'
import { BrowserRouter as Router, Route, Switch, Link } from 'react-router-dom'; 
import { ListPageContext } from "./helpers/appContext";

export const carData = {
    cars: [{"stockNumber":10000,"manufacturerName":"Tesla","modelName":"Model S","color":"silver","mileage":{"number":151884,"unit":"km"},"fuelType":"Diesel","pictureUrl":"https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"}],
    totalPageCount:100,
    totalCarsCount:1000
}
export const isFetched = false;
const setTotalCarsCount = jest.fn();
const setTotalPageCount = jest.fn();
const setPagenumber = jest.fn();
const setSelectedFilter = jest.fn();
const setSavedCars = jest.fn();
const setAvailableFilters = jest.fn();
const setFilterParams = jest.fn();

const pageContextProvider = {totalCarsCount:1,
    setTotalCarsCount,
    totalPageCount:2,
    setTotalPageCount,
    pagenumber:1,
    setPagenumber,
    selectedFilter:[{}],
    setSelectedFilter,
    savedCars:[],
    setSavedCars,
    availableFilters:{test:["test value"]},
    setAvailableFilters,
    filterParams: [{test:"test filter value"}],
    setFilterParams
}

const AllTheProviders = ({ children }) => {
    return (
     <Router>
          <ListPageContext.Provider value={pageContextProvider}>{children}</ListPageContext.Provider>
     </Router>
    )
}

const customRender = (ui, options) => {
    render(ui, { wrapper: AllTheProviders, ...options });
};

export * from "@testing-library/react";
export { customRender as render };