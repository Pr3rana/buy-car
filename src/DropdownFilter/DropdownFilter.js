import './DropdownFilter.css';
import React, { useState, useContext, useEffect } from 'react';
import { ListPageContext } from "../helpers/storeContext";
import AutoSuggestPanel from './AutoSuggestPanel'
import Button from '../Button/Button';

export default function Filter() {
    // const [filtersType] = useState(["Color", "Manufacturers"]);
    const [availableFilters, setAvailableFilters] = useState({});
    // const [availableManufacturersFilters, setAvailableManufacturersFilters] = useState([]);
    const [filterParams, setFilterParams] = useState({});
    const {setPagenumber, setSelectedFilter} = useContext(ListPageContext);
    const ROOT_URL = "https://auto1-mock-server.herokuapp.com/api";
    // const {data: colors} = useFetch(ROOT_URL+"/colors");
    // const {data: manufacturer} = useFetch(ROOT_URL+"/manufacturer");
    const typeRefs = {};
    Object.keys(availableFilters).forEach((type,index)=>{
        typeRefs[type] = React.createRef(null);
        typeRefs[type+"Default"] = React.createRef(null);
    })

    const handleSelect = (e)=>{
        // console.log(e.target)
        const type = e.target.getAttribute("data-filter-type");
        const value = e.target.innerText.toLowerCase();
        setFilterParams({...filterParams, [type.toLowerCase()]:value});
        console.log(e.target.previousSibling, "prev")
        typeRefs[type+"Default"].current.innerText = value;
        typeRefs[type].current.style.display!==""?typeRefs[type].current.style.display="":typeRefs[type].current.style.display="block";
    }

    const handleFilterClick = (e)=>{
        const type = e.target.getAttribute("data-type");
        typeRefs[type].current.style.display!==""?typeRefs[type].current.style.display="":typeRefs[type].current.style.display="block";
    }

    const applyFilter = ()=>{
        console.log(filterParams, "filters")
        let params = ''
        for(let key in filterParams){
            params+=`${key}=${filterParams[key]}&`
        }
        setPagenumber(1);
        setSelectedFilter(params)
    }
    
    useEffect(()=>{
        async function fetchFilterData(){
            let [ colors, manufacturers ]  =  await Promise.all([
                fetch(`${ROOT_URL}/colors`).then(value => value.json()),
                fetch(`${ROOT_URL}/manufacturers`).then(value => value.json())
            ])
            // setAvailableColorFilters(colors.colors);
            const listOfManufacturers = manufacturers.manufacturers.map(key=>key.name)
            // setAvailableManufacturersFilters(listOfManufacturers);
            setAvailableFilters({Color:colors.colors, Manufacturer: listOfManufacturers});
        }
        fetchFilterData()
    },[]);
    
    return ( 
        <div className="filter-wrapper">
            {Object.keys(availableFilters).map((type,index)=>(
                <div className="custom-select" key={type}>
                    <p>{type}</p>
                    <div ref={typeRefs[type+"Default"]} className="default-selected-item" onClick={handleFilterClick} data-type={type}>All {type}</div>
                    <AutoSuggestPanel ref={typeRefs[type]} key={index} filterData={availableFilters[type]} type={type} select={handleSelect}/>
                </div>
            ))}
            <Button value="Filter" click={applyFilter}/>
        </div>
     );
}
