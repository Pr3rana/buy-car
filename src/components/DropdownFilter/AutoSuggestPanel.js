import React from 'react';
import './DropdownFilter.css';

function AutoSuggestPanel ({filterData=[], select, type}, ref) {
    return(
      <div data-testid='autosuggest-panel-wrapper' ref={ref} className="autosuggestPanel">
        {
          filterData.map((el, index) => {
          return (
              <div data-testid ="suggestion-row" className="suggestionRow" key = {el+index} onClick={(e)=>select(e)}  data-filter-type={type}>{el.toLowerCase()}</div>
          )
        })
        }
      </div>
    )
  }
  const typeRef = React.forwardRef(AutoSuggestPanel);
  export default typeRef;