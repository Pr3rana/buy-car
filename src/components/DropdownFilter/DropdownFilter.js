import './DropdownFilter.css';
import React, { useContext, useEffect } from 'react';
import { ListPageContext } from "../../helpers/storeContext";
import AutoSuggestPanel from './AutoSuggestPanel'
import Button from '../Button/Button';

export default function Filter() {
    const {setPagenumber, setSelectedFilter, availableFilters, setAvailableFilters, filterParams, setFilterParams} = useContext(ListPageContext);
    const ROOT_URL = "https://auto1-mock-server.herokuapp.com/api";
    const typeRefs = {};
    Object.keys(availableFilters).forEach((type,index)=>{
        typeRefs[type] = React.createRef(null);
        typeRefs[type+"Default"] = React.createRef(null);
    })

    const handleSelect = (e)=>{
        const type = e.target.getAttribute("data-filter-type");
        const value = e.target.innerText.toLowerCase();
        setFilterParams({...filterParams, [type.toLowerCase()]:value});
        console.log(e.target.previousSibling, "prev")
        typeRefs[type+"Default"].current.innerText = value;
        if(typeRefs[type].current.style.display!==""){
            typeRefs[type].current.style.display="";
            typeRefs[type+"Default"].current.className = "default-selected-item";
        }
        else {
            typeRefs[type].current.style.display="block";
            typeRefs[type+"Default"].current.className+=' select-arrow-active';
        }
    }

    const handleFilterClick = (e)=>{
        const type = e.target.getAttribute("data-type");
        if(typeRefs[type].current.style.display!==""){
            typeRefs[type].current.style.display="";
            typeRefs[type+"Default"].current.className = "default-selected-item";
        }
        else{
            typeRefs[type].current.style.display="block";
            typeRefs[type+"Default"].current.className+=' select-arrow-active';
        }
    }

    const applyFilter = (e)=>{
        if(Object.keys(filterParams).length === 0){
            e.stopPropagation();
            return
        }
        let params = ''
        for(let key in filterParams){
            params+=`${key}=${filterParams[key]}&`
        }
        setPagenumber(1);
        setSelectedFilter(params)
    }
    
    useEffect(()=>{
        async function fetchFilterData(){
            let [ colors, manufacturers ]  =  await Promise.all([
                fetch(`${ROOT_URL}/colors`).then(value => value.json()),
                fetch(`${ROOT_URL}/manufacturers`).then(value => value.json())
            ])
            const listOfManufacturers = manufacturers.manufacturers.map(key=>key.name)
            setAvailableFilters({Color:colors.colors, Manufacturer: listOfManufacturers});
        }
        fetchFilterData()
    },[]);
    
    return ( 
        <div data-testid="filter-wrapper" className="filter-wrapper">
            {Object.keys(availableFilters).length===0?(<div className="filter-loader"><img alt="Loader" src="https://i.gifer.com/ZZ5H.gif" /></div>):
            Object.keys(availableFilters).map((type,index)=>(
                <div className="custom-select" key={type}>
                    <p className="filter-type">{type}</p>
                    <div data-testid="default-filter" ref={typeRefs[type+"Default"]} className="default-selected-item" onClick={handleFilterClick} data-type={type}>All {type}</div>
                    <AutoSuggestPanel ref={typeRefs[type]} key={index} filterData={availableFilters[type]} type={type} select={handleSelect}/>
                </div>
            ))}
            <Button customClass="filter-btn" value="Filter" click={applyFilter}/>
        </div>
     );
}
