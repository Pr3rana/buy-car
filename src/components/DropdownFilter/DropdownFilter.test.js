import { render, screen, cleanup } from '../../test-utils.js';
import DropdownFilter from './DropdownFilter';

afterEach(cleanup)
describe("Filter component render", () => {
  /*DropdownFilter*/
  test("initial renders", () => {
    render(<DropdownFilter />)
    expect(screen.getByTestId('filter-wrapper')).toMatchSnapshot(
      `<div class="custom-select"><p class="filter-type">test</p><div data-testid="default-filter" class="default-selected-item" data-type="test">All test</div>
      <div data-testid="autosuggest-panel-wrapper" class="autosuggestPanel"><div data-testid="suggestion-row" class="suggestionRow" data-filter-type="test">test value</div></div>
      </div><button data-testid="default-btn" class="default-btn filter-btn">Filter</button>`
    )
  });
})

describe("Filter props",()=>{
  test("Test filter with mock data",()=>{
    render(<DropdownFilter />)
    expect(screen.getByTestId('filter-wrapper')).toBeVisible()
    expect(screen.getByTestId('default-filter')).toBeInTheDocument()
    expect(screen.getByTestId('default-filter')).toHaveTextContent("All test")
    expect(screen.queryByTestId('autosuggest-panel-wrapper')).toBeVisible()
    expect(screen.getByTestId('autosuggest-panel-wrapper')).toMatchSnapshot(
      `<div data-testid="autosuggest-panel-wrapper" class="autosuggestPanel">
      <div data-testid="suggestion-row" class="suggestionRow" data-filter-type="test">test value</div></div>`
    )
  })
})