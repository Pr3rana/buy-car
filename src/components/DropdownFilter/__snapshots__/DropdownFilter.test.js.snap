// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component render initial renders: <div class="custom-select"><p class="filter-type">test</p><div data-testid="default-filter" class="default-selected-item" data-type="test">All test</div>
      <div data-testid="autosuggest-panel-wrapper" class="autosuggestPanel"><div data-testid="suggestion-row" class="suggestionRow" data-filter-type="test">test value</div></div>
      </div><button data-testid="default-btn" class="default-btn filter-btn">Filter</button> 1`] = `
<div
  class="filter-wrapper"
  data-testid="filter-wrapper"
>
  <div
    class="custom-select"
  >
    <p
      class="filter-type"
    >
      test
    </p>
    <div
      class="default-selected-item"
      data-testid="default-filter"
      data-type="test"
    >
      All 
      test
    </div>
    <div
      class="autosuggestPanel"
      data-testid="autosuggest-panel-wrapper"
    >
      <div
        class="suggestionRow"
        data-filter-type="test"
        data-testid="suggestion-row"
      >
        test value
      </div>
    </div>
  </div>
  <button
    class="default-btn filter-btn"
    data-testid="default-btn"
  >
    Filter
  </button>
</div>
`;
